src/org/mozilla/javascript/NativeConsole.java

```diff
@@ -216,61 +217,94 @@ public static String format(Context cx, Scriptable scope, Object[] args) {
-                case "%i":
-                    double number = ScriptRuntime.toNumber(args, argIndex);
-                    if (Double.isInfinite(number) || Double.isNaN(number)) {
-                        replaceArg = "0";
-                    } else {
-                        replaceArg = String.valueOf(((Double) number).intValue());
-                    }
-                    break;
-
-                case "%f":
-                    replaceArg = String.valueOf(ScriptRuntime.toNumber(args, argIndex));
-                    break;
 
-                case "%o":
-                case "%O":
-                    replaceArg = formatObj(cx, scope, args, argIndex);
-                    break;
-
-                default:
-                    replaceArg = "";
-                    break;
+            if (placeHolder.equals("%%")) {
+                replaceArg = "%";
+            } else if (argIndex >= args.length) {
+                replaceArg = placeHolder;
+                argIndex++;
+            } else {
+                Object val = args[argIndex];
+                switch (placeHolder) {
+                    case "%s":
+                        replaceArg = ScriptRuntime.toString(val);
+                        break;
+
+                    case "%d":
+                    case "%i":
+                        replaceArg = formatInt(val);
+                        break;
+
+                    case "%f":
+                        replaceArg = formatFloat(val);
+                        break;
+
+                    case "%o":
+                    case "%O":
+                        replaceArg = formatObj(cx, scope, args, argIndex);
+                        break;
+
+                    default:
+                        replaceArg = "";
+                        break;
+                }
+                argIndex++;
             }
 
-            argIndex++;
             matcher.appendReplacement(buffer, Matcher.quoteReplacement(replaceArg));
         }
         matcher.appendTail(buffer);
 
         return buffer.toString();
     }
 
-    private static String formatObj(Context cx, Scriptable scope, Object[] args, int argIndex) {
-        if (argIndex >= args.length) {
-            return "";
+    private static String formatFloat(Object val) {
+        if (val instanceof BigInteger || val instanceof Symbol) {
+            return ScriptRuntime.NaNobj.toString();
+        }
+
+        return String.valueOf(ScriptRuntime.toNumber(val));
+    }
+
+    private static String formatInt(Object val) {
+        if (val instanceof BigInteger) {
+            return ScriptRuntime.bigIntToString((BigInteger) val, 10) + "n";
+        }
+
+        if (val instanceof Symbol) {
+            return ScriptRuntime.NaNobj.toString();
+        }
+
+        double number = ScriptRuntime.toNumber(val);
+
+        if (Double.isInfinite(number) || Double.isNaN(number)) {
+            return ScriptRuntime.toString(number);
         }
 
+        return String.valueOf((int) number);
```

Comment:
`(long)` might be better here.

This PR
```js
// const INT32_MAX = 2147483647

console.log('%d', 2147483647)
// INFO 2147483647

console.log('%d', 2147483648)
// INFO 2147483647
```
Node.js
```js
console.log('%d', 2147483648)
// 2147483648
```
Firefox
```js
console.log('%d', 2147483648)
// -2147483648

console.log('%d', 2147483649)
// -2147483647
```

Reply:
Changed to `long`.